### PR TITLE
feat(project): validate dates before casting

### PR DESCRIPTION
**Pull request by gregoryagu**
Fix so that a date is not created if the value is null.

I fixed this because aurelia-orm uses typer, as it will create a date even when the value is null. So I was getting date values where there should be none.

### CODE REVIEW

**Comment by gregoryagu** [comment]
Forgive me if I have not done the pull request properly.  It gave me a message that the commit message was invalid, so I updated that but it is still showing as error.   (I am very new to doing pull requests)

**Comment by RWOverdijk** [comment]
What value do you expect when there's no date? Cast it to null?

In any case, the commit message format should then be:

`feat(project): validate dates before casting`

**Comment by gregoryagu** [comment]
Yes, cast it to null.  In other words, a null date should stay a null date, and not become a date with the minimum value.

**Comment by RWOverdijk** [comment]
Honestly, I don't think that should be typer's job. Typer just types. If you want a null check, that's validation, not typing. I'm sorry.

**Comment by gregoryagu** [comment]
The point is that a nullable date value becomes a date value.  That is not typing.  In other words, my null date in the database suddenly becomes 1/1/1972 when it is sent to the client.  It is Typers job to assign a date where there was none before?

In other words, Typer is changing a nullable date field, to a date field.  A nullable date is a valid type, and that type should not be changed by Typer.

The reason this is important is that Typer is used by Aurelia-ORM, so every nullable date field that I have in the database ends up with a date.  For example a task can have a start date, and an end date, or neither.  If I use typer, those are always populated, even when the user has not entered anything for them.   

And I don't see any downside to this.  When would you want a date of 1/1/1972 to show up as your date where there was none before?

**Comment by RWOverdijk** [comment]
That's fair, but doing so for only dates is weird. When you think of for example float, you get NaN. It's the behaviour of casting. When you have a default field in for example MySQL, it's `1970-01-01 00:00:01` which also isn't what you expect but allows you to treat it as a date. Using null, means you want it to be a different type. You decided to add a secondary type, null.

So in my opinion, this should really be done by a validator. When you cast null to a date, you get `1970-01-01T00:00:00.000Z`. Doing anything differently introduces magic and I really don't want magic in a tiny library that has one specific job.

Aurelia-orm is a great example. There, you should be handling the defaults when populating the entity. If anything, it's a design flaw in aurelia-orm and it really should be cast to null when it's a date and if that's the behaviour you expect.

Long story short: this module wasn't intended to validate input. But I'm conflicted. I do see your point in simply wanting null to be null for dates. I also agree it's strange that this is the behaviour. But the module is doing precisely what is asked of it and that behaviour shouldn't change.

We could add a check on this line: https://github.com/SpoonX/aurelia-orm/blob/c8a85554d740513d6b73db1d1c982a371fc87405/src/repository.js#L232

We could check if the type is date, and the value is falsy. If so, we can simply keep the value as null. This would be a breaking change, and would need to be documented as such.

This sounds like the most elegant and proper solution to me. Do you agree?

**Comment by gregoryagu** [comment]
Yes, thank you very much! 

I really do appreciate all the fine work on all of the libraries that SpoonX produces.

**Comment by RWOverdijk** [comment]
@gregoryagu Thanks, that's nice to hear!

Would you be willing to make that PR? I can help with the commit message.

**Comment by gregoryagu** [comment]
@RWOverdijk Yes, I would be happy to.

**Comment by gregoryagu** [comment]
@RWOverdijk  Done.  (But I am not sure what I did wrong on the commit message. )
